docs/dev/python/table-api-users-guide/udfs/python_udfs.md

```diff
@@ -227,3 +227,164 @@ def iterable_func(x):
+        accumulator[1] -= weight
+        
+    def get_result_type(self):
+        return DataTypes.BIGINT()
+        
+    def get_accumulator_type(self):
+        return DataTypes.ROW([
+            DataTypes.FIELD("f0", DataTypes.BIGINT()), 
+            DataTypes.FIELD("f1", DataTypes.BIGINT())])
+
+
+env = StreamExecutionEnvironment.get_execution_environment()
+table_env = StreamTableEnvironment.create(env)
+# the result type and accumulator type can also be specified in the udaf decorator:
+# weighted_avg = udaf(WeightedAvg(), result_type=DataTypes.BIGINT(), accumulator_type=...)
+weighted_avg = udaf(WeightedAvg())
+t = table_env.from_elements([(1, 2, "Lee"),
+                             (3, 4, "Jay"),
+                             (5, 6, "Jay"),
+                             (7, 8, "Lee")]).alias("value", "count", "name")
+
+# call function "inline" without registration in Table API
+result = t.group_by(t.name).select(weighted_avg(t.value, t.count).alias("avg")).to_pandas()
+print(result)
+
+# register function
+table_env.create_temporary_function("weighted_avg", WeightedAvg())
+
+# call registered function in Table API
+result = t.group_by(t.name).select(call("weighted_avg", t.value, t.count).alias("avg")).to_pandas()
+print(result)
+
+# register table
+table_env.create_temporary_view("source", t)
+
+# call registered function in SQL
+result = table_env.sql_query(
+    "SELECT weighted_avg(`value`, `count`) AS avg FROM source GROUP BY name").to_pandas()
+print(result)
+{% endhighlight %}
+
+The `accumulate(...)` method of our `WeightedAvg` class takes three inputs. The first one is the accumulator
+and the other two are user-defined inputs. In order to calculate a weighted average value, the accumulator
+needs to store the weighted sum and count of all the data that has been accumulated. In our example, we
+use a `Row` object to be the accumulator. Accumulators are automatically managed
+by Flink's checkpointing mechanism and are restored in case of a failure to ensure exactly-once semantics.
+
+### Mandatory and Optional Methods
+
+**The following methods are mandatory for each `AggregateFunction`:**
+
+- `create_accumulator()`
+- `accumulate(...)` 
+- `get_value(...)`
+
+**The following methods of `AggregateFunction` are required depending on the use case:**
+
+- `retract(...)` is required when there are other operations that generate retract messages before current UDAF call, e.g. group aggregate , outer join. \
+This method is optional, but it is strongly recommended to be implemented to ensure the UDAF can be used in any use case.
+- `get_result_type()` and `get_accumulator_type()` is required if the result type and accumulator type would not be specified in the `udaf` decorator.
+
+### ListView and MapView
+
+If an accumulator needs to store large amounts of data, `pyflink.table.ListView` and `pyflink.table.MapView` 
+provide advanced features for leveraging Flink's state backends in unbounded data scenarios. 
```

Comment:
```suggestion
could be used instead of list and map. These two data structures provide the similar functionalities as list and map, however usually having better performance by leveraging Flink's state backend to eliminate unnecessary state access.
```